File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;
int ans = 0;
vector<vector<int>> duns;
vector<bool> vis;

void dfs(int power, int cnt) {
    ans = max(ans, cnt);
    
    for(int i=0; i<duns.size(); i++) {
        if(!vis[i] && power >= duns[i][0]) {
            vis[i] = true;
            dfs(power-duns[i][1], cnt+1);
            vis[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    
    duns = dungeons;
    vis.resize(duns.size(), false);
    
    dfs(k, 0);
    return ans;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    int answer = -1;
    int k; 
    int[][] dungeons;
    boolean[] vis;
    ArrayList<Integer> picks = new ArrayList<>();
    
    int Explore() {
        int cnt = 0;
        int energy = k;
        
        for(int p : picks) {
            if(dungeons[p][0] <= energy) {
                energy -= dungeons[p][1];
                cnt++;
            }
        }
        return cnt;
    }
    
    void dfs(int cnt) {
        if(cnt == dungeons.length) {
            answer = Math.max(answer, Explore());
        }
        
        for(int i=0; i<dungeons.length; i++) {
            if(!vis[i]) {
                vis[i] = true; picks.add(i);
                dfs(cnt+1);
                vis[i] = false; picks.remove(picks.size()-1);
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        this.k = k;
        this.dungeons = dungeons;
        vis = new boolean[dungeons.length];
        
        dfs(0);
        
        return answer;
    }
}